main/ruletester/ruletester.go

```diff
@@ -72,132 +105,295 @@ func main() {
 	cassandraConfig := cassandra.CassandraMetricMetadataConfig{
 		Hosts:    config.MetricMetadataConfig.Hosts,
 		Keyspace: config.MetricMetadataConfig.Keyspace,
 	}
-	apiInstance := common.NewMetricMetadataAPI(cassandraConfig)
+	_ = common.NewMetricMetadataAPI(cassandraConfig)
 
-	var output *os.File
-	if *unmatchedFile != "" {
-		output, err = os.Create(*unmatchedFile)
-		if err != nil {
-			common.ExitWithMessage(fmt.Sprintf("Error creating the output file: %s", err.Error()))
-		}
-	}
-	stat := run(ruleset, scanner, apiInstance, output)
-	report(stat)
+	// var output *os.File
+	// if *unmatchedFile != "" {
+	// 	output, err = os.Create(*unmatchedFile)
+	// 	if err != nil {
+	// 		common.ExitWithMessage(fmt.Sprintf("Error creating the output file: %s", err.Error()))
+	// 	}
+	// }
+	// stat := run(ruleset, scanner, apiInstance, output)
+	// report(stat)
 }
 
-func run(ruleset util.RuleSet, scanner *bufio.Scanner, apiInstance api.MetricMetadataAPI, unmatched *os.File) Statistics {
+type ChunkResult struct {
+	matched                int
+	unmatched              int
+	reverse_convert_failed int
+}
+
+func DoAnalysis(metrics []string, graphiteConverter util.RuleBasedGraphiteConverter) {
+	graphiteConverter.EnableStats()
+
+	workQueue := make(chan []string, len(metrics)/25+1)
+	resultQueue := make(chan ChunkResult, len(metrics))
+	unmatchedQueue := make(chan string, len(metrics))
 	var wg sync.WaitGroup
-	stat := Statistics{
-		perMetric: make(map[api.MetricKey]PerMetricStatistics),
+
+	i := 0
+	for i = 0; i+25 < len(metrics); i = i + 25 {
+		workQueue <- metrics[i : i+25]
 	}
-	type result struct {
-		input   string
-		result  api.TaggedMetric
-		success bool
+	if i < len(metrics) {
+		workQueue <- metrics[i:]
 	}
-	inputBuffer := make(chan string, 10)
-	outputBuffer := make(chan result, 10)
-	done := make(chan struct{})
-	for id := 0; id < runtime.NumCPU(); id++ {
+
+	close(workQueue)
+
+	fmt.Printf("Starting work...\n")
+	for j := 0; j < 10; j++ {
+		wg.Add(1)
 		go func() {
+			counter := 0
+			defer wg.Done()
 			for {
 				select {
-				case <-done:
-					return
-				case input := <-inputBuffer:
-					metric, matched := ruleset.MatchRule(input)
-					outputBuffer <- result{
-						input,
-						metric,
-						matched,
+				case metrics, more := <-workQueue:
+					counter++
+					if counter%100 == 0 && counter != 0 {
+						fmt.Printf(".")
+					}
+					chunk_result := ChunkResult{}
+					for _, metric := range metrics {
+						graphiteMetric := util.GraphiteMetric(metric)
+						taggedMetric, err := graphiteConverter.ToTaggedName(graphiteMetric)
+						if err != nil {
+							_, err := graphiteConverter.ToGraphiteName(taggedMetric)
+							if err != nil {
+								chunk_result.reverse_convert_failed++
+							} else {
+
+							}
+							chunk_result.matched++
+						} else {
+							unmatchedQueue <- metric
+							chunk_result.unmatched++
+						}
+					}
+					resultQueue <- chunk_result
+
+					if !more {
```

Comment:
Should be at the top of this `case`. If `more` is false, then the value `metrics` will be the zero value (nil), so looping over it is pointless.